packages/leancode_lint/lib/lints/missing_cleanup.dart

```diff
@@ -0,0 +1,374 @@
+    var classNode = node.parent;
+    while (classNode != null && classNode is! ClassDeclaration) {
+      classNode = classNode.parent;
+    }
+    return switch (classNode) {
+      final ClassDeclaration classNode => classNode,
+      _ => null,
+    };
+  }
+
+  String? _getDisposableMethodName(InterfaceType type) {
+    return type.methods2
+            .firstWhereOrNull(
+              (method) => config.cleanupMethods.contains(method.name3),
+            )
+            ?.name3 ??
+        type.element3.inheritedMembers.entries
+            .firstWhereOrNull(
+              (entry) =>
+                  config.cleanupMethods.contains(entry.key.name) &&
+                  entry.value.baseElement is MethodElement2,
+            )
+            ?.key
+            .name;
+  }
+
+  bool _isWidgetType(InterfaceType type) {
+    const widgetTypeChecker = TypeChecker.fromName(
+      'Widget',
+      packageName: 'flutter',
+    );
+
+    return widgetTypeChecker.isExactlyType(type) ||
+        widgetTypeChecker.isSuperTypeOf(type);
+  }
+
+  bool _isInReturnWidget(InstanceCreationExpression node) {
+    AstNode? currentNode = node;
+    var returnsWidget = false;
+    var isInReturn = false;
+
+    bool isCurrentNodeReturn() =>
+        currentNode is ReturnStatement || currentNode is ExpressionFunctionBody;
+
+    while (currentNode != null && !(returnsWidget && isInReturn)) {
+      if (!isInReturn && isCurrentNodeReturn()) {
+        isInReturn = true;
+      }
+      if (currentNode case MethodDeclaration(
+        returnType: NamedType(:final InterfaceType type),
+      ) when _isWidgetType(type)) {
+        returnsWidget = true;
+      }
+
+      currentNode = currentNode.parent;
+    }
+    return returnsWidget && isInReturn;
+  }
+
+  bool _isStateOfWidget(ClassDeclaration classNode) =>
+      switch (classNode.declaredElement) {
+        final element? => const TypeChecker.fromName(
+          'State',
+          packageName: 'flutter',
+        ).isAssignableFrom(element),
+        _ => false,
+      };
+
+  bool _isWidgetClass(ClassDeclaration classNode) =>
+      switch (classNode.declaredElement) {
+        final element? => const TypeChecker.fromName(
+          'Widget',
+          packageName: 'flutter',
+        ).isAssignableFrom(element),
+        _ => false,
+      };
+}
+
+class _DisposeExpressionsGatherer extends GeneralizingAstVisitor<void> {
+  _DisposeExpressionsGatherer({required this.targetName});
+
+  final String targetName;
+
+  final List<InvocationExpression> _disposeExpressions = [];
+
+  static List<InvocationExpression> gatherForTarget({
+    required AstNode node,
+    required String targetName,
+  }) {
+    final visitor = _DisposeExpressionsGatherer(targetName: targetName);
+    node.accept(visitor);
+    return visitor._disposeExpressions;
+  }
+
+  @override
+  void visitExpressionStatement(ExpressionStatement node) {
+    if (node.expression
+        case MethodInvocation(
```

Comment:
Since you only care about expressions that are MethodInvocations, you can just override the `visitMethodInvocation` method of the visitor